### PR TITLE
Add block asset documentation

### PR DESCRIPTION
**Pull request by mikejolley**
Adds documentation covering the AbstractBlock class and JIT asset registration.

### CODE REVIEW

File: docs/contributors/block-assets.md [hunk 1]
```diff
@@ -1 +1,56 @@
+# Block Script Assets
 
+[Block Types](https://github.com/woocommerce/woocommerce-gutenberg-products-block/tree/master/src/BlockTypes) are often responsible for enqueing script assets that make blocks functional on both the front-end and within the editor. Additionally, some block scripts require extra data from the server and thus have extra dependencies that need to be loaded.
+
+For performance reasons the blocks plugin ensures assets and asset data is made available only as needed.
+
+## When are assets needed?
+
+Assets are needed when we know a block will being rendered.
```
Comment (nerrad) [inline docs/contributors/block-assets.md:9]: Typo:

> Assets are needed when we know a block will be rendered.

File: docs/contributors/block-assets.md [hunk 2]
```diff
@@ -1 +1,56 @@
...
+
+## When are assets needed?
+
+Assets are needed when we know a block will being rendered. 
+
+In the context of [Block Types](https://github.com/woocommerce/woocommerce-gutenberg-products-block/tree/master/src/BlockTypes), assets and asset data is enqueued within the block `render()` method.
+
+In an admin editor context we must also ensure asset *data* is available when the `enqueue_block_editor_assets` hook is fired. That is because block scripts are enqueued ready for the Block Inserter, but the block may not be rendered. 
+
+Note: `enqueue_block_editor_assets` fires regardless of whether or not a block has been rendered in the editor context, so unless handled correctly, block data may be loaded twice.
```
Comment (nerrad) [inline docs/contributors/block-assets.md:15]: Should mention be made here about _how_ to handle this correctly or that using the provided api takes care of that?

**Comment by senadir** [comment]
what is the objective of this doc, is it to document the methods of that class or is it to explain the reasoning and benefit of this method and how to implement it in a block?, after reading this, I still struggle to understand it and have a vague understanding of things works, I might be to go to each method and reference this doc, but if that's the case, I think we might be better off documenting the class directly.

**Comment by mikejolley** [comment]
@senadir documenting a specific part of implementing new blocks as part of this repo—assets need specific handling and unless you are aware of how things work in editor vs frontend it can be hard to get right.

**Pull request merged**